Support zero-byte match in `countMatches` function

ahmadov opened the pull request:
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Previously, function `countMatches` would stop counting at the first empty match even if the pattern accepts it.
To overcome this issue, `countMatches` now continues execution by advancing by a single character when an empty match occurs. Users who like to retain the old behavior can enable setting `count_matches_stop_at_empty_match`.

### Documentation entry for user-facing changes
- [ ] Documentation is written (mandatory for new features)

ahmadov committed:
Support zero-byte match in `countMatches` function

alexey-milovidov commented:
This is not a bug fix.
The category should be a "backward incompatible change".

alexey-milovidov commented:
> countmatches_zero_match_legacy_behavior

Name it: count_matches_stop_at_empty_match

ahmadov committed:
support legacy behavior (stop at empty match) via a new settings variable

ahmadov committed:
update docs

ahmadov committed:
update docs and settings change history

rschu1ze commented on src/Core/SettingsChangesHistory.cpp:
```diff
@@ -81,6 +81,7 @@ const VersionToSettingsChangesMap & getSettingsChangesHistory()
             {"use_skip_indexes_if_final", 0, 1, "Change in default value of setting"},
             {"use_skip_indexes_if_final_exact_mode", 0, 1, "Change in default value of setting"},
             {"min_outstreams_per_resize_after_split", 0, 24, "New setting."},
+            {"count_matches_stop_at_empty_match", false, false, "New setting."},
```
See l. 66-67. This needs to be

```cpp
{"count_matches_stop_at_empty_match", true, false, "New setting."},
```

so that users with [the compatibility setting](https://clickhouse.com/docs/operations/settings/settings#compatibility) configured to something pre-25.6 will get the previous behavior.

ahmadov committed:
Handle randomized settings for `countMatches`

ahmadov committed:
Enable `count_matches_stop_at_empty_match` for performance tests

Since the new behavior might be slow since it consumes all input instead
of stopping at the first empty match.

ahmadov committed:
update countMatches performance test settings

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/countMatches-zero-byte-match

ahmadov committed:
update version information

rschu1ze commented on src/Core/SettingsChangesHistory.cpp:
```diff
@@ -82,6 +82,7 @@ const VersionToSettingsChangesMap & getSettingsChangesHistory()
             {"use_skip_indexes_if_final_exact_mode", 0, 1, "Change in default value of setting"},
             {"allow_experimental_time_series_aggregate_functions", false, false, "New setting to enable experimental timeSeries* aggregate functions."},
             {"min_outstreams_per_resize_after_split", 0, 24, "New setting."},
+            {"count_matches_stop_at_empty_match", true, false, "New setting."},
```
Oops, the new setting is still in the 25.6 section. The 25.7 setting would be the right one.

Two options to fix this:
- in a small separate PR
- add the `pr-must-backport` label and bring this PR into 25.6 (and only this version).
